### PR TITLE
Add restrictions for user created invites

### LINKED ISSUE #6353: [Feature] restrictions on user generated invites
When allowing users to generate invites, it'd be nice to be able to restrict them to a certain number of uses.

At the moment it seems that regular users don't get the "Never expires" option, so their invites are only good for up to 24 hours, but they can still create any number of infinite use invites, as often as they want.

I'd like to be able to limit the uses they can specify for any given invite (to, say 1 or 5), and limit the number of invites they can have active at any one time (like, 1 or 2).

Like a lot of new admins, I'm needing to keep very tight control over the growth of my instance, while I figure out the resource usage and cost.

* * * *

- [ ] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [ ] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

### PR DESCRIPTION
**Pull request by Plastikmensch**
Ported from https://github.com/glitch-soc/mastodon/pull/2229
Closes #6353

Currently, the only way to prevent the invite feature from being abused is to prevent users from creating them entirely.
This adds a middle ground which would allow users to create invites, but with restrictions:
1. An account has to be at least one week old to be able to create invites.
2. A user can only have 5 non-expired invites at a time.
3. A user can only create 20 invites in the past 24h.
4. User created invites are limited to single use.

These limits can be bypassed by granting the `bypass_invite_limits` permission, which is given to the moderator and admin role by default.

The following env vars have been added:
`MAX_ACTIVE_INVITES` - The amount of active invites a user can create
`MAX_DAILY_INVITES` - The amount of total invites a user can create in 24h
`MIN_DAYS_BEFORE_INVITE` - The amount of days an account has to exist before creating invites

Some considerations:
For large instances, this could significantly increase the amount of created invites. How much of an issue this is in practice, I can't say.
For instances with approval only registrations, afaik the approval date is not saved, so should the approval take longer than a week, the account could be able to create invites right away.

### CODE REVIEW

File: app/models/user_role.rb
Comment (ClearlyClaire) [inline app/models/user_role.rb:39]: I think instead of adding a `bypass_invite_limits` permission, I would add a `restricted_invites` permission or similar, thus preserving the current meaning and not requiring a database migration.

File: app/policies/invite_policy.rb
Comment (ClearlyClaire) [inline app/policies/invite_policy.rb:9]: It would make sense for `7.days` to be a constant, if not configurable. Assuming such a constant would be called `MIN_TIME_BEFORE_INVITE`, a slightly more idiomatic way to write it would be `current_user.created_at <= MIN_TIME_BEFORE_INVITE.ago`

File: app/validators/invite_validator.rb
Comment (ClearlyClaire) [inline app/validators/invite_validator.rb:21]:
```suggestion
    invite.user.invites.where(created_at: 1.day.ago...).count >= DAILY_LIMIT
```

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
Hi!

Thank you for your contribution! Having more granularity regarding allowing users to send invites is definitely a welcome change.

I wonder if some of the limits make sense, or if they should be configurable.

I would also flip the new permission around to not require a migration.

**Comment by Wild1145** [comment]
Could we have all of these as configurable toggles? Rather than the entire thing being on/off or having to give bypass permissions out it would be nice that I could enable say 1 or 2 of the 4. 

(It might already be the case, I'm not a code person but just saw the PR description)

File: app/models/user_role.rb
Comment (Plastikmensch) [reply app/models/user_role.rb:39]: Thank you for reviewing this!

The idea behind this was since invites are enabled by default, restrictions should be applied by default too.
And as staff should be exempt, this unfortunately necessitates a migration.
Though in theory, I could invert this and add it to the default permission, but I think that would also affect staff.
So to clarify: Should this kept as a default or not?

File: app/policies/invite_policy.rb
Comment (Plastikmensch) [reply app/policies/invite_policy.rb:9]: Should that constant be in this file, or added to the validator?

File: app/validators/invite_validator.rb
Comment (Plastikmensch) [reply app/validators/invite_validator.rb]: Thanks!

**Comment by Plastikmensch** [comment]
> Could we have all of these as configurable toggles? Rather than the entire thing being on/off or having to give bypass permissions out it would be nice that I could enable say 1 or 2 of the 4.
> 
> (It might already be the case, I'm not a code person but just saw the PR description)

That wouldn't make much sense, because these could easily bypassed then.
Let's say, you only enable the active limit. Now, a malicious user could just use invites as they are created.
Let's say, you enable both the active limit and daily limit. Now a malicious user could just invite themself and create more invites.
One restriction doesn't accomplish much without the others.

File: app/models/user_role.rb
Comment (ClearlyClaire) [reply app/models/user_role.rb:39]: Oh, you are right, I forgot invites were handled as “default permissions” and that it was not explicitly set for moderators and admins but inherited from the default…

However, existing installs will already have a “Everyone” role that explicitly includes the invite permission, so no migration is required for them.

So I think a way forward that does not involve running a migration is the following:
- make the new permission flag be `restricted_invite_users` or something instead of `bypass_invite_limits` and have `invite_users` keep its existing scope
- change the default flags to be `restricted_invite_users` rather than `invite_users`
- add `invite_users` to `config/roles.yml`
- change `validate_dangerous_permissions` to allow both `invite_users` and `restricted_invite_users` (with e.g. a `Flags::UNPRIVILEGED` constant that would be a superset of `Flags::DEFAULT`)

**Comment by ClearlyClaire** [comment]
> That wouldn't make much sense, because these could easily bypassed then.
> Let's say, you only enable the active limit. Now, a malicious user could just use invites as they are created.

True!

> Let's say, you enable both the active limit and daily limit. Now a malicious user could just invite themself and create more invites.

An admin could give the permission to a special role rather than to Everyone, in which case the age limit would be replaced by a manual approval.

There is also an opportunity to have a setting to have invites not bypass approvals, which could also change the dynamic of that.

But I think both of those things can be addressed after this PR is merged.

**Comment by Wild1145** [comment]
> That wouldn't make much sense, because these could easily bypassed then.
> Let's say, you only enable the active limit. Now, a malicious user could just use invites as they are created.
> Let's say, you enable both the active limit and daily limit. Now a malicious user could just invite themself and create more invites.
> One restriction doesn't accomplish much without the others.

I think it's more the point where I as a server admin want to have control over invites, currently we have 0 control beyond enabling / disabling them, and this doesn't really give us a great deal of control either. 

As a real world example, when I was struggling to scale I wanted my users to be able to invite others and have it auto approved, many wouldn't have been a week old but I would also have liked to have been able to limit the total number of invites a user could create to stem the number of people joining, yes those new people could have created invites, but it was about slowing the tide of users rather than preventing malicious users, which in reality aren't the type of person this is going to stop anyway... 

All of these limits can already be bypassed by a hostile user and would be if they genuinely wanted to bypass them, I just want as an admin to have actual control over my invites rather than the code only catering for one type of server on the network.

File: app/models/user_role.rb
Comment (Plastikmensch) [reply app/models/user_role.rb:39]: Sorry I haven't addressed any of these requested changes yet.
It's more involved than I'm happy with, but I'm currently working on it and it's mostly done, but I haven't tested it yet and I still need to rewrite the specs.
I'm not comfortable pushing unfinished work and I really need to restrict myself on how much I do, so I unfortunately can't give a timeframe.

Comment (Plastikmensch) [reply app/models/user_role.rb:39]: > The owner role is easy to fix by removing `restricted_invite_users` from `ALL`, but I wasn't able to find a save way to do that with the admin and moderator role.

Well, I didn't account for that roles can't give permissions they don't have themselves. So either the owner role can't create invites without limits being applied, or they can't set that permission for other roles.
Or I make yet another exemption.
At this point I'm basically refactoring how roles work, so honestly I would rather leave this as is.

File: db/post_migrate/20230522141920_add_bypass_invite_limits_permission.rb
Comment (Plastikmensch) [inline db/post_migrate/20230522141920_add_bypass_invite_limits_permission.rb:21]: This migration would fail should `bypass_invite_limits` be removed at any future point.

```suggestion
  class UserRole < ApplicationRecord
    FLAGS = {
      bypass_invite_limits: (1 << 20),
    }
   end

  def change
    admin_role     = UserRole.find_by(name: 'Admin')
    moderator_role = UserRole.find_by(name: 'Moderator')

    if admin_role
      admin_role.permissions |= UserRole::FLAGS[:bypass_invite_limits]
      admin_role.save
    end

    return unless moderator_role

    moderator_role.permissions |= UserRole::FLAGS[:bypass_invite_limits]
    moderator_role.save
  end
```

**Pull request closed without merging**
